### ISSUE TITLE
Have Table Functions announce range of output row counts

### ISSUE THREAD

**Issue by frankmcsherry** [MaterializeInc/materialize#8150]
Our table functions are allowed to produce multiple output columns, which means that they are occasionally the way we apply 1:1 functions that produce multiple columns. The information that a table function is 1:1 is valuable, as it means that the optimizer can for example remove table functions whose columns are not used, or for which we just need counts, that sort of thing. Instead, at the moment we are conservative and imagine that all table functions might produce varying numbers of rows, and so cannot be removed.

There is a follow-up task to use this information, where `Demand` and projection pushdown would be the two places to start, but this is just the first task of adding that information.

**Comment by frankmcsherry** [comment]
For prior art, Rust iterators have such a method that returns `(usize, Option<usize>)`, indicating a lower bound (which can safely be set to zero) and an optional upper bound if it is known.

**Comment by xudong963** [comment]
New to Materialize, I want to try the task

**Comment by lluki** [comment]
Is this issue about rows or about columns? 

I think we have a pretty good handle of *columns* already: the `output_type` method of `TableFunc` reports the `RelationType` (column count with types) of this table function.  `output_arity` reports the number of columns. Similar methods are defined on `MirRelationExpr`. 

We dont know much about *rows*. It is possible to implement a `size_hint` with the suggested `(usize, Option<usize>)` return type.  This has to be defined on `MirRelationExpr`, otherwise we don't know about the input to the table function. I think there are two cases: One where the input is a constant - in this case the optimizer is already able to eliminate the functions (at least it is in the cases i could come up with). In the other case, the input is unknown at optimization time, thus the `size_hint` would always return `(0, None)` which won't allow any new optimizations either.

An option might be to return a row-multiplier, for example we know that `Map` does not change the number of rows and then would return `0`. But even here I can't see a case where this helps with non-constant inputs.

Am I missing something?

**Comment by frankmcsherry** [comment]
Yes, I think the intent is about announcing "the multiplier": not the total number of rows, but the number of rows that a `TableFunc` would emit on *each input*.

We should maybe sniff around for a specific success case. The one I was thinking of was #9147 but that has been addressed more directly. The "general" pattern of usefulness though are optimizations that struggle to move around `FlatMap` stages because of its cardinality-modifying behavior. We could take a look at these and try to find some candidates, but I agree that this is an important next step before going too deep.

From the issue: 

> There is a follow-up task to use this information, where Demand and projection pushdown would be the two places to start, but this is just the first task of adding that information.

This is very oblique, so we might have to work to reconstruct these. 
1. Projection pushdown is an optimization that says "if you aren't using a column, you may not need to use its inputs columns either". It is foiled by `TableFunc`s at the moment, because even if you don't look at its output, the columns on which it depends could influence the *number* of output rows. If we are lucky and have a `TableFunc` that has a certain number of output rows (say one) for each input row, then we can update the projection pushdown optimization to say "don't mark input columns just because the `TableFunc` references them".
2. `Demand` is an archaic form of the projection pushdown optimization. 

If you check out `projection_pushdown.rs`, and sniff around for the `FlatMap` case, you find:
```rust
    MirRelationExpr::FlatMap { input, func, exprs } => {
        let inner_arity = input.arity();
        // A FlatMap which returns zero rows acts like a filter
        // so we always need to execute it
        let mut columns_to_pushdown =
            desired_projection.iter().cloned().collect::<BTreeSet<_>>();
        for expr in exprs.iter() {
            columns_to_pushdown.extend(expr.support());
        }
        columns_to_pushdown.retain(|c| *c < inner_arity);

        reverse_permute(exprs.iter_mut(), columns_to_pushdown.iter());
        let columns_to_pushdown = columns_to_pushdown.into_iter().collect::<Vec<_>>();
        self.action(input, &columns_to_pushdown, gets);
        // The actual projection always has the newly-created columns at
        // the end.
        let mut actual_projection = columns_to_pushdown;
        for c in 0..func.output_type().arity() {
            actual_projection.push(inner_arity + c);
        }
        actual_projection
    }
```
which you can see always includes the `expr.support()` for each expression. Fixing that would be one use of the announced output row count range, if it were tight.

Having said all of that, I think we do want one of @wangandi or @aalexandrov to have an example query they care about, before sinking much time. I think *this issue*, without the follow-on, could be pretty quick in either case, but if it does nothing because we didn't have an example where it would do a think, that would be too bad!

**Comment by lluki** [comment]
Thanks, it's much clearer now. I'll try to come up with a query first

**Comment by lluki** [comment]
I found one that could benefit from this information.
```
materialize=> EXPLAIN SELECT f1 FROM (SELECT f1,generate_series(f1,f1) FROM t1);
                              Optimized Plan
---------------------------------------------------------------------------
 Source materialize.public.t1 (u41):                                      +
 | Project (#0, #1)                                                       +
                                                                          +
 Query:                                                                   +
 %0 =                                                                     +
 | Get materialize.public.t1 (u41)                                        +
 | FlatMap generate_series(#0, #0, 1)                                     +
 | Reduce group=(#0, #1)                                                  +
 | | agg row_number(record_create(list_create(record_create(#0, #1, #2))))+
 | Project (#2)                                                           +
 | FlatMap unnest_list(#0)                                                +
 | Map record_get[0](record_get[1](#1))                                   +
 | Project (#2)                                                           +

(1 row)
```

**Comment by lluki** [comment]
We can give some non-trivial bounds for:

 * `regexp_extract` produces at most no. of match group rows. This can be deduced from the regex, which has a good chance of being static.
 * `csv_extract` either produces 0 or 1 row, depending on if the input data has the right amount of columns.

It's not clear (yet) how this information can be used to optimize queries further.

**Comment by lluki** [comment]
As discussed with Ufuk, we don't work on this unless we find better queries that could profit from this optimization.
